File: abc256/e.cpp
#include <bits/stdc++.h>
#define ll long long
using namespace std;

class UnionFind
{

private:
    vector<long long> parent_or_size;
    long long size;

public:
    UnionFind(long long s)
    {
        size = s;
        parent_or_size.assign(size, -1);
    }

    long long root(long long x)
    {
        if (parent_or_size[x] < 0)
        {
            return x;
        }
        parent_or_size[x] = root(parent_or_size[x]);
        return parent_or_size[x];
    }
    long long is_same(long long x, long long y)
    {
        return root(x) == root(y);
    }
    void unite(long long x, long long y)
    {
        long long rx = root(x);
        long long ry = root(y);
        if (rx == ry)
        {
            return;
        }
        parent_or_size[rx] += parent_or_size[ry];
        parent_or_size[ry] = rx;
    }
};

int main()
{
    int N;
    cin >> N;
    vector<ll> X(N), C(N);
    for (auto &x : X)
    {
        cin >> x, --x;
    }
    for (auto &c : C)
    {
        cin >> c;
    }

    auto uf = UnionFind(N);
    ll ans = 0;
    for (int i = 0; i < N; i++)
    {
        // 嫌っている相手と同じ閉路に所属しているかどうか？
        if (!uf.is_same(i, X[i]))
        {
            uf.unite(i, X[i]);
            continue;
        }
        // 嫌っている相手と同じ閉路に所属している。→ 閉路内で最も少ない不満度を足していく
        auto v = i;
        ll cur = 1e10;
        do
        {
            cur = min(C[v], cur);
            v = X[v];
        } while (v != i);
        ans += cur;
    }
    cout << ans << endl;
}

File: abc256/e.py
import sys
sys.setrecursionlimit(2*10**5 + 1)

class UnionFind:
    def __init__(self,N):
        self.__parents_or_size = [-1 for i in range(N)]
    
    def root(self,x):
        if self.__parents_or_size[x] < 0:
            return x
        else:
            self.__parents_or_size[x] = self.root(self.__parents_or_size[x])
            return self.__parents_or_size[x]
    
    def same(self,x,y):
        return self.root(x) == self.root(y)
    
    def unite(self,x,y):
        x = self.root(x)
        y = self.root(y)
        if x == y:
            return
        self.__parents_or_size[x] += self.__parents_or_size[y]
        self.__parents_or_size[y] = x

N = int(input())
X = list(map(int,input().split()))
C = list(map(int,input().split()))
X = [x-1 for x in X]
directed_graph = [[] for i in range(N)]

for i,x in enumerate(X):
    directed_graph[i].append(x-1)

uf = UnionFind(N+1)
ans = 0
for i in range(N):
    if not uf.same(i, X[i]):
        uf.unite(i, X[i])
    else:
        n = i
        min_stress = 1e10
        while True:
            min_stress = min(min_stress,C[n])
            n = X[n]
            if n == i:
                break
        ans += min_stress

print(ans)